File: packages/client/src/async.js
/* Launch MicropedeClients asynchronously */
const _ = require('lodash');
const uuidv1 = require('uuid/v1');
const uuidv4 = require('uuid/v4');
const {MicropedeClient, GenerateClientId} = require('./client.js');
const DEFAULT_TIMEOUT = 5000;

const CLIENT_OPTIONS = {resubscribe: false};

class MicropedeAsync {
  constructor(appName, host="localhost", port=undefined, version='0.0.0') {
    if (appName == undefined) throw "appName undefined";
    const name = `micropede-async-${uuidv1()}-${uuidv4()}`;
    try {
      this.client = new MicropedeClient(appName, host, port, name, version, CLIENT_OPTIONS);
      this.client.listen = _.noop;
    } catch (e) {
      console.error(this.dumpStack(this.name, e));
    }
  }
  async reset() {
    /* Reset the state of the client (use between actions)*/

    // Generate a new clientId (so that each sub is easier to debug)
    let {host, port, name, appName} = this.client;
    this.client.clientId = GenerateClientId(name, appName);
    try {
      // Disconnect and Reconnect the MicropedeClient for this async instance
      await this.client.disconnectClient();
      await this.client.connectClient(this.client.clientId, host, port);
    } catch (e) {
      throw e;
    }
  }

  async getState(sender, prop, timeout=DEFAULT_TIMEOUT) {
    /* Get the state of another plugins property */
    const label = `${this.client.appName}::getState`;
    const topic = `${this.client.appName}/${sender}/state/${prop}`;
    let done = false;
    let timer;

    try {
      this.enforceSingleSubscription(label);
      await this.reset();
    } catch (e) {
      throw(this.dumpStack([label, topic], e));
    }

    // Subscribe to a state channel of another plugin, and return
    // the first response
    return new Promise((resolve, reject) => {

      // Success case: (receive message from state channel)
      this.client.onStateMsg(sender, prop, (payload, params) => {
        if (timer) clearTimeout(timer);
        done = true;
        this.client.disconnectClient().then((d) => {
          resolve(payload);
        }).catch((e) => {
          reject(e);
        });
      });

      // Rejection case: (client times out before receiving state msg)
      timer = setTimeout( () => {
        if (!done) {
          done = true;
          this.client.disconnectClient().then((d) => {
            reject([label, topic, `timeout ${timeout}ms`]);
          }).catch((e) => {
            reject([label, topic, `timeout ${timeout}ms`]);
          });
        }
      }, timeout);

    });
  }

  async getSchema(receiver, timeout=DEFAULT_TIMEOUT) {
    /* Get the schema used by the plugin */
    const payload = await this.triggerPlugin(receiver, "get-schema", {}, timeout);
    return payload.response;
  }

  async getSubscriptions(receiver, timeout=DEFAULT_TIMEOUT) {
    /* Get the subscriptions of another plugin */
    const payload = await this.triggerPlugin(receiver, "get-subscriptions", {}, timeout);
    return payload.response;
  }

  async putPlugin(receiver, property, val, timeout=DEFAULT_TIMEOUT) {
    /* Call put on another plugin */
    // Wrap string payloads into objects (since the put endpoint expects headers)
    if (!_.isPlainObject(val)) {
      let msg = {}; _.set(msg, property, val);
      val = msg;
    }
    // Call a put action on the receiving plugin
    const result = await this.callAction(receiver, property, val, "put",
          timeout);
    return result;
  }

  async triggerPlugin(receiver, action, val={}, timeout=DEFAULT_TIMEOUT) {
    /* Call trigger on another plugin */
    const result = await this.callAction(receiver, action, val,
      "trigger", timeout);
    return result;
  }

  async callAction(receiver, action, val, msgType='trigger', timeout=DEFAULT_TIMEOUT) {
    /* Call action (either trigger or put) and await notification */
    const label = `${this.client.appName}::callAction::${msgType}::${action}`;
    let done = false;
    let timer;

    // Remove the timeout if set to -1 (some actions may not notify immediately)
    let noTimeout = (timeout == -1) ? true : false;

    // Setup header
    _.set(val, "__head__.plugin_name", this.client.name);
    _.set(val, "__head__.version", this.client.version);

    // Create a mqtt topic based on type, receiver, and action
    const topic = `${this.client.appName}/${msgType}/${receiver}/${action}`;

    // Reset the state of the MicropedeAsync client
    try {
      this.enforceSingleSubscription(label);
      await this.reset();
    } catch (e) {
      throw(this.dumpStack([label, topic], e));
    }

    // Await for notifiaton from the receiving plugin
    return new Promise((resolve, reject) => {

      this.client.onNotifyMsg(receiver, action, (payload, params) => {
        if (timer) clearTimeout(timer);
        done = true;
        this.client.disconnectClient().then((d) => {
          if (payload.status) {
            if (payload.status != 'success') {
              reject(_.flattenDeep([label, _.get(payload, 'response')]));
              return;
            }
          } else {
            console.warn([label, "message did not contain status"]);
          }
          resolve(payload);
        }).catch((e)=>{
          reject(e)
        });
      });

      this.client.sendMessage(topic, val);

      // Cause the notification to fail after given timeout
      if (!noTimeout) {
        timer = setTimeout( () => {
          if (!done) {
            done = true;
            this.client.disconnectClient().then((d) => {
              reject([label, topic, `timeout ${timeout}ms`]);
            }).catch((e) => {
              reject([label, topic, `timeout ${timeout}ms`]);
            });
          }
        }, timeout);
      }

    });

  }

  dumpStack(label, err) {
    /* Dump stack between plugins (technique to join stack of multiple processes') */
    if (!err) return _.flattenDeep([label, 'unknown error']);
    if (err.stack)
      return _.flattenDeep([label, JSON.stringify(err.stack).replace(/\\/g, "").replace(/"/g,"").split("\n")]);
    if (!err.stack)
      return _.flattenDeep([label, JSON.stringify(err).replace(/\\/g, "").replace(/"/g,"").split(",")]);
  }

  enforceSingleSubscription(label) {
    /* Ensure that MicropedeAsync instances are only handling one sub at a time */
    const totalSubscriptions = this.client.subscriptions.length;
    const defaultSubscriptions = this.client.defaultSubCount;
    if (totalSubscriptions - defaultSubscriptions > 1 ) {
      const msg = 'only one active sub per async client';
      throw(this.dumpStack([label, msg]));
    }
  }

}

module.exports = MicropedeAsync;

File: packages/python/micropede/async.py
import asyncio
import json
import uuid

import pydash as _

from .client import MicropedeClient, generate_client_id, set_timeout

DEFAULT_TIMEOUT = 5000

class MicropedeAsync():
    def __init__(self, app_name, host='localhost', port=None, version='0.0.0', loop=None):
        try:
            if (app_name is None):
                raise(Exception("app_name is None"))
            name = f'micropede-async-{uuid.uuid1()}-{uuid.uuid4()}'
            self.client = MicropedeClient(app_name, host=host, port=port,
                                          name=name, version=version, loop=loop)
            self.safe = self.client.safe
            self.client.listen = _.noop
        except Exception as e:
            raise Exception(self.dump_stack(self.client.name, e))

    async def reset(self):
        host = self.client.host
        port = self.client.port
        name = self.client.name
        app_name = self.client.app_name
        self.client.client_id = generate_client_id(name, app_name)
        await self.client.disconnect_client()
        await self.client.connect_client(self.client.client_id, host, port)
        return

    async def get_state(self, sender, prop, timeout=DEFAULT_TIMEOUT):
        label = f'{self.client.app_name}::get_state'
        topic = f'{self.client.app_name}/{sender}/state/{prop}'
        timer = None
        future = asyncio.Future(loop=self.client.loop)

        try:
            self.enforce_single_subscription(label)
            await self.reset()
        except Exception as e:
            raise Exception(self.dump_stack(self.client.name, e))

        def on_state_msg(payload, params):
            # clear timeout ??
            def on_disconnect(d):
                if (future.done() == False):
                    future.set_result(payload)
            f1 = self.client.disconnect_client()
            f1.add_done_callback(self.safe(on_disconnect))

        self.client.on_state_msg(sender, prop, on_state_msg)
        def on_timeout():
            if future.done():
                return
            def on_disconnect(d):
                future.set_exception(
                    self.dump_stack(label, [topic, f'timeout {timeout}ms']))
            f1 = self.client.disconnect_client()
            f1.add_done_callback(self.safe(on_disconnect))

        set_timeout(self.safe(on_timeout), timeout)

        return await future

    async def get_subscriptions(self, receiver, timeout=DEFAULT_TIMEOUT):
        payload = await self.trigger_plugin(receiver, 'get-subscriptions', {}, timeout)
        return _.get(payload, 'response')

    async def put_plugin(self, receiver, prop, val, timeout=DEFAULT_TIMEOUT):
        if (not _.is_dict(val)):
            msg = {}
            _.set_(msg, prop, val)
            val = msg
        result = await self.call_action(receiver, prop, val,  'put', timeout)
        return result

    async def trigger_plugin(self, receiver, action, val={}, timeout=DEFAULT_TIMEOUT):
        result = await self.call_action(receiver, action, val, 'trigger', timeout)
        return result

    async def call_action(self, receiver, action, val={}, msg_type='trigger',
                             timeout=DEFAULT_TIMEOUT):
        label = f'{self.client.app_name}::callAction::{msg_type}::{action}'
        future = asyncio.Future(loop=self.client.loop)
        # import pdb; pdb.set_trace()

        done = False
        timer = None
        no_timeout = False
        if timeout is -1:
            no_timeout = True

        _.set_(val, '__head__.plugin_name', self.client.name)
        _.set_(val, '__head__.version', self.client.version)
        topic = f'{self.client.app_name}/{msg_type}/{receiver}/{action}'
        try:
            self.enforce_single_subscription(label)
            await self.reset()
        except Exception as e:
            raise self.dump_stack(label, [topic, e])

        def on_notify(payload, params):
            # clear timer??
            def on_disconnect(d):
                if future.done():
                    return

                if (_.get(payload, 'status') is not None):
                    if (_.get(payload, 'status') != 'success'):
                        future.set_exception(
                            self.dump_stack(label, _.get(payload, 'status')))
                else:
                    print("WARNING: ", label, 'message did not contain status')

                if future.done() == False:
                    future.set_result(payload)

            f1 = self.client.disconnect_client()
            f1.add_done_callback(self.safe(on_disconnect))

        self.client.on_notify_msg(receiver, action,
                                  self.safe(on_notify))

        self.client.send_message(topic, val)

        if (no_timeout is not None):
            def on_timeout():
                if future.done():
                    return
                def on_disconnect(d):
                    future.set_exception(
                        self.dump_stack(label, [topic, f'timeout {timeout}ms']))
                f1 = self.client.disconnect_client()
                f1.add_done_callback(self.safe(on_disconnect))
            set_timeout(self.safe(on_timeout), timeout)

        return await future

    def enforce_single_subscription(self, label):
        total_subscriptions = len(self.client.subscriptions)
        if hasattr(self.client, 'default_sub_count'):
            default_subscriptions = self.client.default_sub_count
        else:
            # TODO: This is a temporary fix
            # Should wait for micropede client to set default_sub_count
            default_subscriptions = 2

        if (total_subscriptions - default_subscriptions > 1):
            msg = 'only one active sub per async client'
            raise self.dump_stack(label, msg)

    def dump_stack(self, label, err):
        if (err is None):
            return Exception(_.flatten_deep([label, 'unknown error']))
        if (_.get(err, 'args')):
            return Exception(_.flatten_deep([label, err.args]))
        else:
            return Exception(_.flatten_deep([label, err]))
